### ISSUE TITLE
Trajectory-optimization solver recompiles every solve, leaks mmaps, and cannot share its jit cache safely

### ISSUE THREAD

**Issue by iory** [iory/scikit-robot#871]
`AugmentedLagrangianSolver` rebuilds and recompiles its cost functions on every `solve()`. The compiled executables are never reused, they leak mmap regions until the process dies, and the cache that exists cannot be shared safely because the traced functions close over the problem's FK arrays.

## The cache never hits, and could not safely be made to

`_build_functions` closes over `fk_data` (link translations, joint axes, base pose, collision offsets) and returns freshly `jax.jit`-wrapped closures. `jax.jit` keys on function identity, so every `solve()` compiles from scratch.

`_jit_cache` exists but is an instance attribute, and callers that build a solver per solve never hit it. Making it module-level would not be safe as it stands: neither `get_problem_structure_key` nor `get_problem_value_hash` includes any FK quantity, so two different robots with the same joint count, the same residual set and the same targets produce the same key. A shared cache would hand robot A's kinematics to robot B and return a plausible, wrong trajectory. `_viser.py` already caches solvers per `(robot_id, group_name)`, which happens to keep FK constant, but nothing enforces that.

## Measurements

29-joint chain, 2 waypoints, 4 box obstacles, 300 inner iterations x 20 outer. Timing the `inner_loop` calls separately (the first carries compilation, the rest do not):

```
call  1   2.436 s   with compilation
call  2   6.505 s   with compilation
call  3-19  ~1.55 s each
```

Of 41.0 s for one solve: compilation about 5.8 s (14%), execution about 29.5 s (72%). Re-solving an identical problem in the same process costs 39.1 s against the first solve's 42.5 s, i.e. nothing is reused.

Each solve also leaks roughly 870 mmap regions that are never released. With the default `vm.max_map_count` of 65530 this reaches the limit after about 65 solves and the process dies with

```
LLVM compilation error: Cannot allocate memory
```

while RAM is still free. Anything that solves many problems in one process -- a design search, a sampling-based study, batch evaluation -- hits this. `jax.clear_caches()` releases the regions exactly, and costs nothing today precisely because nothing is being reused, but it would be the wrong fix once reuse works.

## Proposal

Pass the values as runtime arguments and keep only the structure in the closure. Joint count, joint types and array shapes decide the traced graph and must stay constant; link transforms, joint axes, base pose, collision offsets and joint limits are values and can be arguments.

Then one compiled executable serves every problem whose structure matches, the cache can be shared safely because the FK is no longer baked in, and the mmap growth stops on its own.

It also makes `jax.vmap` over problems possible, which is the real payoff for anyone solving many similar problems. I have a working prototype of this shape. Against the sequential path it is bit-identical -- `max|dq| = 0`, equal costs, equal success flags -- with the outer loop's two Python branches (break on convergence, raise the penalty) replaced by per-candidate masks so a converged problem is frozen rather than iterated further.

Speedups, measured:

```
synthetic benchmark, 72 problems, CPU     7.9x
synthetic benchmark, 72 problems, GPU    58x
real workload, batch width 8              5.6x
real workload, batch width 4              1.7x
```

Batch width is bounded by how many problems share an FK array shape, which is narrower than the problem count -- in the real workload 32 candidates collapsed to groups of at most 8 -- so the synthetic figures are a ceiling, not a forecast. Worth noting that a single problem is 4.4x *slower* on GPU than on CPU; the GPU only pays off batched.

Happy to send this as a PR: four commits, each bit-identical, with the batched entry point last, plus a test that pins the batched and sequential results together.

**Comment by iory** [comment]
Fixed by #872. The FK arrays now travel as runtime arguments, so a compiled executable serves every problem of one structure, and `solve_batched` widens the inner loop with `jax.vmap` for callers that sweep a family of them.

**Comment by iory** [comment]
Reopening: #872 made the fix *possible* but did not land it. I closed this
too early.

What #872 did deliver: the FK arrays travel as a runtime argument, so a
compiled executable no longer bakes in one problem's kinematics. That is the
precondition this issue said had to hold before any cache could be shared.
It also added `solve_batched` for callers that can group their problems.

What is still true on main:

- `_jit_cache` is an instance attribute (`self._jit_cache = {}` in
  `__init__`), and `solve_ik` calls `create_solver` on every invocation, so
  the cache starts empty every time and never hits. A caller that solves
  many problems of one structure still compiles once per problem.
- The mmap regions still accumulate for the same reason. With the default
  `vm.max_map_count` of 65530 a long-running sweep still reaches the limit
  and dies with `LLVM compilation error: Cannot allocate memory` while RAM
  is free.

Now that FK is an argument, the cache can be moved to module scope safely --
which was the part that was previously unsound, since neither
`_get_problem_structure_key` nor `_get_problem_value_hash` includes any FK
quantity. The remaining work is that move, plus deciding what else must
enter the key now that FK no longer needs to (residual set, waypoint count,
obstacle counts, prismatic mask, array shapes).

Worth roughly 14% of a solve on the workload measured above -- compilation
was 5.8 s of 41 s -- but the real payoff is that the mmap growth stops, so
a sweep no longer has to call `jax.clear_caches()` between problems to
survive.
